### PR TITLE
Refactor rematerialization logic into Operation and support more modes

### LINKED ISSUE #23106: 'Operation' object has no attribute 'rematerialized_call' when using RematScope
When a keras.Operation (or a subclass that is not a Layer) is executed within a RematScope, it crashes with an AttributeError. This is because the Operation.__call__ implementation attempts to call self.rematerialized_call, but that method is currently only defined in the Layer subclass, not in the base Operation class.

Reproduction script: https://colab.research.google.com/drive/1PuIB2GfMbGX087n7epQM6m-CSOXRVE_2?usp=sharing

### PR DESCRIPTION
**Pull request by buildwithsuhana**
Fixes https://github.com/keras-team/keras/issues/23106

## Description
This PR fixes a critical AttributeError and functional gap in the rematerialization system by migrating advanced gradient checkpointing logic from the Layer subclass to the base Operation class. Previously, executing a raw keras.Operation within a RematScope would crash because Operation.__call__ attempted to invoke rematerialized_call, a method that was only fully implemented for Layer objects. By unifying this logic in the base class, this change ensures that all Keras operations—not just layers—correctly respect memory optimization modes like larger_than and list_of_layers, providing a consistent and robust memory-saving strategy across the entire functional and sequential API ecosystem. Consistent with this refactor, the Layer class now only handles the activations mode locally while delegating all other rematerialization decisions to the Operation parent, and the test suite has been expanded to verify standalone operation rematerialization.

## Contributor Agreement

**Please review our [AI-Assisted Contribution Policy](../CONTRIBUTING.md#ai-assisted-contribution-policy) and check all boxes below before submitting your PR for review:**

- [x] I am a human, and not a bot.
- [x] I will be responsible for responding to review comments in a timely manner.
- [x] I will work with the maintainers to push this PR forward until submission.

**Note:** Failing to adhere to this agreement may result in your future PRs no longer being reviewed.

### CODE REVIEW

File: keras/src/layers/layer.py [hunk 1]
```diff
@@ -1769,36 +1767,8 @@ def rematerialized_call(self, layer_call, *args, **kwargs):
...
-                tree.flatten(tree.map_structure(compute_size, output_spec))
-            )
-            if (
-                output_size
-                and output_size > self._remat_mode.output_size_threshold
-            ):
-                return remat.remat(layer_call)
-        elif self._remat_mode.mode == "activations":
+        remat_mode = getattr(self, "_remat_mode", None)
+        if remat_mode is not None and remat_mode.mode == "activations":
```
Comment (hertschuh) [inline keras/src/layers/layer.py]: It is safe (and more readable) to just do:
```python
if self._remat_mode == "activations"
```
Because it's initialized in the `__init__`.

File: keras/src/ops/operation.py
Comment (hertschuh) [inline keras/src/ops/operation.py]: In this file, you can now replace all `getattr(self, "_remat_mode", None)` with `self._remat_mode`, this is more readable.

File: keras/src/ops/operation.py [hunk 2]
```diff
@@ -96,6 +100,48 @@ def call(self, *args, **kwargs):
...
+            )
+
+        # Full rematerialization
+        if self._remat_mode.mode == "full":
+            return remat.remat(fn)
+
+        # Apply rematerialization to specific layers/operations
+        elif self._remat_mode.mode == "list_of_layers" and (
+            self.name in self._remat_mode.layer_names
+        ):
```
Comment (hertschuh) [inline keras/src/ops/operation.py]: Hmm... so now it's more of a "list_of_operations". The issue is that we don't provide a way to set the operation name, so this one is not useable for operations.

Can you change it back to only apply to layers?

**Review by hertschuh — CHANGES_REQUESTED** [review]
Thanks for the fix!

**Review by hertschuh — APPROVED** [review]
Thanks for the fix!
